[team6-BE] Jung, yeon 1차 PR

kimnayeon0108 opened the pull request:
안녕하세요 백엔드 Jung, yeon 입니다.
URL 요청에 따른 기본적인 기능 구현까지 완료했습니다. 
배포와 OAuth 구현은 아직 진행하지 않았습니다.
[API 문서](https://github.com/rla36/sidedish/wiki/API-%EB%AC%B8%EC%84%9C)

### 구현사항
- response를 위한 별도의 DishDTO를 만들었습니다. 리스트로 보이는 부분을 위한 MainPageDishDTO와, 상세보기 페이지를 위한 DetailDishDTO를 생성했습니다.
- data를 json 형식으로 DB에 저장하기 위한 RequestDishDTO를 만들었습니다. DetailDishDTO와 동일한 property들에, 추가적으로 categoryID를 가져서 dish들을 저장할 때 원하는 카테고리에 저장할 수 있도록 했습니다.

### 궁금했던 점
- RequestDishDTO 객체를 생성하는 메소드(createDish)를 Dish, DishDTO 중 어느 클래스에 넣을지 고민했습니다. 일단 Entity는 서비스, DTO와 별개로 존재해야된다는 점에서 DTO 클래스에 넣는것이 맞을 것 같아 그렇게 했습니다.

kimnayeon0108 committed:
feat: set spring boot project

kimnayeon0108 committed:
refactor: change directory

rla36 committed:
feat: add mySql dependencies

kimnayeon0108 committed:
feat: complete to connect to mysql

rla36 committed:
docs: changed API document

kimnayeon0108 committed:
feat: add Category entity and enum CategoryType

kimnayeon0108 committed:
feat: add Dish entity (#7)

kimnayeon0108 committed:
feat: add CategoryController (in progress)

kimnayeon0108 committed:
docs: update API

kimnayeon0108 committed:
feat: update schema.sql

kimnayeon0108 committed:
feat: add ConvertUtils to convert String to List type

kimnayeon0108 committed:
feat: add DetailDishDTO for detail page

kimnayeon0108 committed:
feat: add DetailDishDTO for detail page

kimnayeon0108 committed:
feat: add CategoryDTO

kimnayeon0108 committed:
feat: add data.sql and update schema.sql to insert dish date

kimnayeon0108 committed:
feat: update controller to return DTO

kimnayeon0108 committed:
feat: add custom exception, CategoryNotFoundException

kimnayeon0108 committed:
feat: update property name

kimnayeon0108 committed:
feat: add DishController

kimnayeon0108 committed:
feat: add logger in CategoryController

kimnayeon0108 committed:
feat: add Dish constructor of which parameter is DTO

kimnayeon0108 committed:
feat: add custom exception, DishNotFoundException

kimnayeon0108 committed:
feat: add DishRepository

kimnayeon0108 committed:
feat: add constructor with no parameter

kimnayeon0108 committed:
feat: add convert list to String function

kimnayeon0108 committed:
feat: update sql files

rla36 committed:
feat: modified PostMapping method to receive DTOs as list

rla36 committed:
feat: add @JsonProperty annotation to format column names

kimnayeon0108 committed:
feat: add RequestDishDTO

kimnayeon0108 committed:
feat: add CategoryService and move business logic to CategoryService

kimnayeon0108 committed:
feat: add DishService and move business logic to DishService

kimnayeon0108 committed:
feat: delete useless enum

kimnayeon0108 committed:
feat: change data structure of dishes Set to List

kimnayeon0108 committed:
feat: override equals and hashCode

kimnayeon0108 committed:
feat: change data structure of dishes Set to List

kimnayeon0108 committed:
chore

kimnayeon0108 committed:
add: return empty list when string is empty

kimnayeon0108 committed:
fix: extend length of detail_images

kimnayeon0108 committed:
feat: add getting access to GitHub by getting access token

ksundong commented on BE/.gitignore:
```diff
@@ -0,0 +1,38 @@
+HELP.md
+.idea/
+.gradle
+build/
+!gradle/wrapper/gradle-wrapper.jar
+!**/src/main/**/build/
+!**/src/test/**/build/
+
```
gitignore가 제대로 적용이 되지 않았는지 .idea/ 에 존재하는 파일들이 버전관리 되고 있네요.
제거해주시는 편이 좋고, 항상 버전관리 되지 않아야 하는 파일들이 커밋에 포함되는지 확인해주세요.

ksundong replied:
또,  추가로 환경별로 제외될 수 있는 파일이 있을 수 있는데요.
gitignore.io에 가보시면 여러 템플릿이 있으니, 적용해보시는 것도 추천드립니다.

rla36 replied:
BE 디렉토리 바깥에 있는 gitignore파일에 제거해야할 파일들을 추가해줬는데, 그래도 .idea폴더가 사라지지 않았습니다.
연이 열심히 검색하셔서 캐시 문제라는 것을 알아내셔서 삭제에 성공했습니다! 
[출처](https://jojoldu.tistory.com/307)에서 시키는대로 하니까 되더라고요 ㅎㅎ

ksundong replied:
고생하셨네요 ㅎㅎ

ksundong commented on BE/README.md:
```diff
@@ -0,0 +1,112 @@
+
+## 메인페이지
+
+- /banchan-code/main
```
`banchan-code/main` 보다는 `sidedishes/main` 이 더 좋지 않을까 하는 생각이 드네요.

ksundong replied:
API는 화면의 구성을 따라가도 되지만, 따라가지 않아도 됩니다.
API를 사용하는건 일반 사용자가 아니니까, API 사용자 관점에서 디자인해보세요.

rla36 replied:
banchan-code가 원래 저희가 합의한 서비스 이름이어서 url에 담았었습니다. 
그런데 말씀해주신대로 서비스이름을 API url설계에 담을 필요가 없더라고요. 
IOS분들과 합의한 내용이 있어서 이번에는 이렇게 진행하고 앞으로 말씀해주신대로 적용해보겠습니다!

ksundong commented on BE/README.md:
```diff
@@ -0,0 +1,112 @@
+    "name": "[미노리키친] 규동 250g",
+    "top_image": "http://public.codesquad.kr/jk/storeapp/data/detail/HBDEF/4cce011a4a352c22cd399a60271b4921.jpg",
+    "description": "일본인의 소울푸드! 한국인도 좋아하는 소고기덮밥",
+    "prices": [6500, 5200],
```
배열말고, 객체로 주는건 어땠을까요?

kimnayeon0108 replied:
@ksundong DB설계가 아쉽다고 하셔서  Badge, ThumbImage, Price, DetailImage 클래스를 만들고 테이블도 생성해서 Dish와 관계를 맺도록 했습니다! 그리고 지금 API를 변경하면 내일 데모때 영향이 갈거 같아서.... 다음부터는 객체로 반환하도록 API 설계 해보겠습니다!

ksundong commented on BE/src/main/java/sidedish/controller/CategoryController.java:
```diff
@@ -0,0 +1,26 @@
+import sidedish.service.dto.CategoryDTO;
+
+@RestController
+@RequestMapping("/banchan-code")
```
위에서도 말씀드렸듯, 더 좋은 Path를 고민해주세요.
개인적으로 한글을 그대로 음으로 가져다 쓰는 것은 좋지 않다고 생각됩니다.

rla36 replied:
위에서 말씀드렸듯이 banchan-code는 저희의 서비스 이름입니다..ㅎㅎ naver.com에서 naver 부분이고 이걸 api 설계의 url에 담을 필요가 없는데 잘못생각했네요!

ksundong commented on BE/src/main/java/sidedish/controller/CategoryController.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    @GetMapping("/{title}")
+    public ResponseEntity<CategoryDTO> mainPage(@PathVariable String title) {
```
`ResponseEntity`를 사용해주신 이유는 무엇인가요?

kimnayeon0108 replied:
@ksundong 사실.... 이전 미션에서 계속 이렇게 해와서 ResponseEntity를 사용한거 같습니다. 이부분 더 학습이 필요할듯 합니다

rla36 replied:
이전 미션에서 계속 사용해왔었는데, header, body, status를 전부 반환해줄 수 있다고 해서 쓰고 있었습니다!
아직 헤더에 무언가를 담아본적은 없지만, ok 리스폰스와 함께 DTO를 반환하려고 사용하고 있었습니다.

사실 다른 대체재가 무엇이 있는지 잘 몰라 계속 이것을 사용한 것 같은데, 추가적인 공부가 필요할 것 같네요..ㅎㅎ

ksundong replied:
Q&A 할 때 잠깐 API를 만들었었는데 그걸 잘 떠올려보세요.

rla36 replied:
CategoryDTO, DishDTO를 리턴하도록 수정했습니다!

ksundong commented on BE/src/main/java/sidedish/controller/CategoryController.java:
```diff
@@ -0,0 +1,26 @@
+
+    @GetMapping("/{title}")
+    public ResponseEntity<CategoryDTO> mainPage(@PathVariable String title) {
+        CategoryDTO categoryDTO = categoryService.convertToCategoryDTO(title);
```
`convertToCategoryDTO`는 별로 좋지 않은 이름으로 보여져요.

kimnayeon0108 replied:
@ksundong createCategoryDTO 라고 변경했습니다! 혹시 이건 괜찮은 이름인지... 궁금합니당

ksundong replied:
저라면, 행위에 집중할 것 같아요.
우리가 하려는건 mainPage를 불러오는 거잖아요?

kimnayeon0108 replied:
@ksundong showMainPage...? 근데 mainPage를 위한 categoryDTO를 얻는다는 의미로 showMainPage보다는 getCategoryDTOForMainPage 가 나은거 같은데 메소드명이 너무 긴 듯 합니다..

ksundong replied:
길어도 명확한 의미를 전달하는 것이 낫다고 생각해요.

ksundong commented on BE/src/main/java/sidedish/controller/DishController.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/banchan-code")
```
CategoryController랑 Mapping이 겹치네요.
사용상에 주의를 요하는 코드는 별로 좋지 않다고 생각됩니다.

ksundong commented on BE/src/main/java/sidedish/controller/DishController.java:
```diff
@@ -0,0 +1,31 @@
+        return ResponseEntity.ok(detailDishDTO);
+    }
+
+    @PostMapping("/dish")
+    public void createDish(@RequestBody List<RequestDishDTO> dishDTOs) {
```
여기도 조금 어색하죠?
dishes가 맞을 것 같고, banchan-code가 아닌 sidedishes였다면 자연스럽게 되었을 것 같네요.

ksundong commented on BE/src/main/java/sidedish/controller/HomeController.java:
```diff
@@ -0,0 +1,21 @@
+@Controller
+public class HomeController {
+
+    @GetMapping("/banchan-code")
+    public String login() {
+        return "redirect:https://github.com/login/oauth/authorize?client_id=0723ab0365a963dc62d7";
+    }
+
+    @GetMapping("/callback")
+    public String getCode(@RequestParam String code) {
+        return "redirect:https://github.com/login/oauth/access_token?client_id=0723ab0365a963dc62d7" +
+                "&client_secret=2739ad01c2c5f84515af7f719fc9fa74921c100e&code=" + code;
+    }
```
여긴 아직 개발중인거죠?
REST API에서 redirect를 시키는 방법은 300번 대 Redirect Code를 주고 Location 헤더에 값을 담아주는 겁니다.

kimnayeon0108 replied:
@ksundong OAuth 구현을 시도해보았습니다. access token을 브라우저에서 발급받으면 이걸 서버에 어떻게 넘겨주지 찾아보다가 Http request를 자바 코드로 보내는 법을 검색해봤습니다. 해당 작업은 어디서 해줘야될지 모르겠어서 다른 백엔드 분들에게 물어봐가면서 학습중입니다.

ksundong commented on BE/src/main/java/sidedish/domain/Dish.java:
```diff
@@ -0,0 +1,132 @@
+    private String thumbImages;
+    private String detailImages;
+
+    public Dish() { }
```
기본 생성자를 정의해주신 이유가 있나요?

kimnayeon0108 replied:
@ksundong 사용하지 않는 기본생성자는 삭제해주었습니다~

ksundong commented on BE/src/main/java/sidedish/domain/Dish.java:
```diff
@@ -0,0 +1,132 @@
+    private String topImage;
+    private String description;
+    private String prices;
+    private String badges;
```
String으로 저장되는것 맞을까요?

kimnayeon0108 replied:
@ksundong 위에서 설명한것과 같이 설계를 변경하여 List<Price> 과 같이 List로 Price, Badge의 객체를 저장하도록 했습니다.

ksundong commented on BE/src/main/java/sidedish/domain/Dish.java:
```diff
@@ -0,0 +1,132 @@
+    private Integer point;
+    private String deliveryInfo;
+    private String thumbImages;
+    private String detailImages;
```
여기도 String으로 저장되는 것이 맞을까요?

ksundong replied:
전체적으로 필드가 많은데, 필드를 연관성이 높은것 끼리 클래스로 모으는 것도 고려해봐주세요.

ksundong commented on BE/src/main/java/sidedish/domain/Dish.java:
```diff
@@ -0,0 +1,132 @@
+
+    public Dish() { }
+
+    public Dish(String name, String topImage, String description, String prices,
+                String badges, Long stock, String deliveryInfo, String thumbImages, String detailImages) {
+        this.name = name;
+        this.topImage = topImage;
+        this.description = description;
+        this.prices = prices;
+        this.badges = badges;
+        this.stock = stock;
+        this.point = createPoint(prices);
+        this.deliveryInfo = deliveryInfo;
+        this.thumbImages = thumbImages;
+        this.detailImages = detailImages;
+    }
```
이렇게 필드가 많으면 생성자로 생성할 때, 헷갈릴 수 있습니다.
빌더패턴의 도입을 고려해보세요~

kimnayeon0108 replied:
@ksundong DishBuilder라는 클래스를 생성해서 빌더패턴을 도입해보면서 든 생각인데, 다른 DTO클래스에 대해서도 Builder 패턴 도입이 필요하지 않나? 라는 생각이 들었습니다. 근데 DTO에 도입을 하면 코드가 더 복잡해지지 않을까 하는 생각이 듭니다ㅠㅠ 이부분은 어떻게 생각하시나요?

ksundong replied:
그래서 Lombok 같은 것을 쓰는것이죠.

ksundong commented on BE/src/main/java/sidedish/domain/Dish.java:
```diff
@@ -0,0 +1,132 @@
+        this.deliveryInfo = dto.getDeliveryInfo();
+        this.thumbImages = ConvertUtils.convertToString(dto.getThumbImages());
+        this.detailImages = ConvertUtils.convertToString(dto.getDetailImages());
+    }
```
Domain 객체에서 DTO를 알고있지 않도록 구현해주세요.
DTO에서 Domain 객체로 변환하는 메서드를 선언해주세요.

kimnayeon0108 replied:
@ksundong 네엡! RequestDTO에서 createDish 라는 메소드를 생성해주었습니다.

ksundong replied:
이것이 변경되는 거니까 convertTo 혹은 to~로 시작하는 이름은 어떨까요?
그냥 의견이에요~

kimnayeon0108 replied:
@ksundong 오호... toDish 좋은거 같습니다!

ksundong commented on BE/src/main/java/sidedish/domain/Dish.java:
```diff
@@ -0,0 +1,132 @@
+    private Integer createPoint(String prices) {
+        String[] priceArr = prices.split(",");
+        return Integer.parseInt(priceArr[0].trim()) / 100;
+    }
```
Dish가 관리해야하는 일일까요?
객체를 분리하면 좋을 것 같네요.

rla36 replied:
사용하지 않는 메소드라 삭제했습니다!

ksundong commented on BE/src/main/java/sidedish/exception/CategoryNotFoundException.java:
```diff
@@ -0,0 +1,8 @@
+package sidedish.exception;
+
+public class CategoryNotFoundException extends RuntimeException{
```
항상 리포맷 신경써주세요~

rla36 replied:
넵넵 띄어쓰기!

kimnayeon0108 replied:
@ksundong 네엡!

ksundong commented on BE/src/main/java/sidedish/exception/DishNotFoundException.java:
```diff
@@ -0,0 +1,8 @@
+package sidedish.exception;
+
+public class DishNotFoundException extends RuntimeException{
```
`XXXNotFoundException`이 여러개 등장하는데, `NotFoundException`을 선언해주는 것은 어떨까요?

rla36 replied:
`NotFoundException`선언하고, DishNotFoundException과 CategoryNotFoundException이 이것을 상속받게 했습니다. 이렇게 하니 익셉션 핸들러에서 두 예외를 한번에 처리할 수 있는 장점이 있었습니다

ksundong commented on BE/src/main/java/sidedish/exception/CategoryNotFoundException.java:
```diff
@@ -0,0 +1,8 @@
+public class CategoryNotFoundException extends RuntimeException{
+
+    public CategoryNotFoundException() {
+        super("해당 카테고리가 존재하지 않습니다!");
```
이런 리터럴도 상수로 추출해보는건 어떨까요?

kimnayeon0108 replied:
@ksundong 상수로 따로 정의해서 사용하도록 변경했습니다!

ksundong commented on BE/src/main/java/sidedish/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+    Optional<Category> findCategoryByTitle(String title);
+
+    Optional<Category> findById(Long id);
```
이건 이미 정의되어 있지 않나요?

kimnayeon0108 replied:
@ksundong 넵.... 굳이 또 명시해줄 필요가 없는거같아 삭제했습니다!

ksundong replied:
기본으로 정의된 것이 어떤것인지 파악해보는 시간도 가져보세요~

ksundong commented on BE/src/main/java/sidedish/service/CategoryService.java:
```diff
@@ -0,0 +1,37 @@
+        this.categoryRepository = categoryRepository;
+    }
+
+    public Category findById(Long id) {
+        Category category = categoryRepository.findById(id).orElseThrow(CategoryNotFoundException::new);
+        return category;
+    }
+
+    public Category findByTitle(String title) {
+        Category category = categoryRepository.findCategoryByTitle(title)
+                .orElseThrow(CategoryNotFoundException::new);
+        return category;
+    }
+
+    public Category save(Category category) {
+        return categoryRepository.save(category);
+    }
```
Category 타입을 리턴해주기보다는 DTO를 전달해주는 편이 좋겠네요~

kimnayeon0108 replied:
@ksundong dishService에서 categoryService대신 repository를 주입받는 걸로 변경해서, save메소드는 필요없게 되어서 삭제했습니다!

ksundong commented on BE/src/main/java/sidedish/service/CategoryService.java:
```diff
@@ -0,0 +1,37 @@
+        return categoryRepository.save(category);
+    }
+
+    public CategoryDTO convertToCategoryDTO(String title) {
+        Category category = findByTitle(title);
+        return new CategoryDTO(category);
+    }
```
위의 피드백을 반영한다면 이 코드는 없어도 되겠죠~?

kimnayeon0108 replied:
@ksundong 위의 내용에 이어서 이 메소드는 createCategoryDTO라고 메소드명을 바꾸고, 이 안에서 repository.save 동작까지 실행하도록 변경했습니다. createCategoryDTO라는 메소드명이 적절한건지... 잘 모르겠습니다.

ksundong commented on BE/src/main/java/sidedish/service/ConvertUtils.java:
```diff
@@ -0,0 +1,39 @@
+
+public class ConvertUtils {
+
+    private ConvertUtils() { }
```
👍

ksundong commented on BE/src/main/java/sidedish/service/ConvertUtils.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.List;
+import java.util.StringJoiner;
+
+public class ConvertUtils {
```
무엇을 Convert하는 건가요?

kimnayeon0108 replied:
@ksundong 어떤걸 convert하는지 명시하기위해서 TypeConvertUtils로 클래스명 변경했습니다.

ksundong commented on BE/src/main/java/sidedish/service/ConvertUtils.java:
```diff
@@ -0,0 +1,39 @@
+    private ConvertUtils() { }
+
+    public static List<Integer> convertToIntegerList(String string) {
+        String[] strArr = string.split(",");
```
축약어 사용은 하지 않았으면 좋겠어요!
구분자도 상수로 꺼내면 좋겠네요.

kimnayeon0108 replied:
@ksundong DB와 객체 설계를 변경하면서 해당 메소드는 사용하지 않게 되어 삭제했습니다. 축약어 사용하지 않기, 상수로 추출하기 기억하겠습니당

ksundong commented on BE/src/main/java/sidedish/service/ConvertUtils.java:
```diff
@@ -0,0 +1,39 @@
+    public static List<Integer> convertToIntegerList(String string) {
+        String[] strArr = string.split(",");
+        List<Integer> integerList = new ArrayList<>();
+        for (String s : strArr) {
```
s 말고 다른 이름 지어주세요~
아니면 stream을 사용해보는건 어떨까요?

kimnayeon0108 replied:
@ksundong 해당 메소드도 위와 같은 이유로 삭제하게 되었습니다. TypeConvertUtils 클래스에 있는 메소드들을 말씀해주신거처럼 stream을 이용해서 구현하니 더 간결하게 구현할 수 있게 된거 같습니다.

ksundong commented on BE/src/main/java/sidedish/service/ConvertUtils.java:
```diff
@@ -0,0 +1,39 @@
+        String[] strArr = string.split(",");
+        List<Integer> integerList = new ArrayList<>();
+        for (String s : strArr) {
+            integerList.add(Integer.parseInt(s.trim()));
```
NumberFormatException이 발생하지 않도록 하려면, 사전에 Validation 해주면 되겠죠.
정규 표현식을 사용해보세요.

kimnayeon0108 replied:
@ksundong 처음에 피드백주신대로 구현하다가 DB 설계 변경하면서 사용하지 않게 되어서 이부분도 삭제했습니다!

ksundong commented on BE/src/main/java/sidedish/service/ConvertUtils.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public static List<String> convertToStringList(String string) {
+        if(string.isEmpty()) {
```
```suggestion
        if (string.isEmpty()) {
```

ksundong commented on BE/src/main/java/sidedish/service/ConvertUtils.java:
```diff
@@ -0,0 +1,39 @@
+        for (String s : strArr) {
+            StringList.add(s.trim());
+        }
+        return StringList;
```
여기도 stream으로 개선해보기 좋겠네요.

ksundong commented on BE/src/main/java/sidedish/service/ConvertUtils.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public static String convertToString(List list) {
+        StringJoiner stringJoiner = new StringJoiner(", ");
+        for (Object object : list) {
+            stringJoiner.add(object + "");
+        }
+        return stringJoiner.toString();
```
Stream의 joining 이용해보세요~

kimnayeon0108 replied:
@ksundong 이부분도 더이상 사용하지 않게 되어 삭제했습니다!

ksundong commented on BE/src/main/java/sidedish/service/DishService.java:
```diff
@@ -0,0 +1,50 @@
+public class DishService {
+
+    private final DishRepository dishRepository;
+    private final CategoryService categoryService;
```
제가 드렸던 피드백 관련해서 이렇게 구현해주신 것 같아요.
내부적으로 얘기를 해봤는데, Service의 계층구조가 생기지 않는 한, Repository를 이용하는 것이 좋다는 의견을 받았습니다.
Repository를 사용해주셔도 좋을 것 같아요.

kimnayeon0108 replied:
@ksundong 네엡! repository를 주입받도록 변경했습니다!

ksundong commented on BE/src/main/java/sidedish/service/DishService.java:
```diff
@@ -0,0 +1,50 @@
+    public DetailDishDTO convertToDetailDishDTO(String title, Long id) {
+        Category category = categoryService.findByTitle(title);
+        Dish dish = findById(id);
+        isDishInCategory(category, dish);
```
category에게 물어봅시다~ dish가 너한테 있니?

kimnayeon0108 replied:
@ksundong category에 있는 hasDish메소드를 호출하는 것으로 변경했습니다!

ksundong commented on BE/src/main/java/sidedish/service/DishService.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public void saveDishes(List<RequestDishDTO> dishDTOs) {
+        for (RequestDishDTO dishDTO : dishDTOs) {
+            Category category = categoryService.findById(dishDTO.getCategoryId());
+            Dish dish = dishDTO.createDish();
+            category.addDish(dish);
+            categoryService.save(category);
+        }
```
변환후에 addAll로 한 번에 넣어주는 것도 고려해봄이 좋겠네요.

kimnayeon0108 replied:
@ksundong 이부분 피드백 반영해서 변경하려고 고민해봤는데 dish마다 다른 category(main, soup, side)에 dish를 저장해야해서 변환 후에 한번에 addAll하는 방법을 생각해내지 못했습니다ㅠ_ㅠ 더 고민해보겠습니다..

ksundong replied:
어렵겠네요 ㅎㅎ

ksundong commented on BE/src/main/java/sidedish/service/dto/CategoryDTO.java:
```diff
@@ -0,0 +1,36 @@
+        this.dishes = convertToMainPageDishDTO(category);
+    }
+
+    private List<MainPageDishDTO> convertToMainPageDishDTO(Category category) {
+        List<Dish> dishes = category.getDishes();
+        List<MainPageDishDTO> mainPageDishDTOs = new ArrayList<>();
+        for (Dish dish : dishes) {
+            mainPageDishDTOs.add(new MainPageDishDTO(dish));
+        }
+        return mainPageDishDTOs;
+    }
```
stream을 이용하면 메서드로 굳이 빼주지 않아도 되겠죠?

kimnayeon0108 replied:
@ksundong 해당 메소드는 삭제하고 생성자 안에서 stream을 이용해서 List<MainPageDishDTO>로 변환하도록 변경했습니다!

ksundong commented on BE/src/main/java/sidedish/service/dto/DetailDishDTO.java:
```diff
@@ -0,0 +1,110 @@
+import sidedish.domain.Dish;
+import sidedish.service.ConvertUtils;
+
+import java.util.List;
+@JsonPropertyOrder({"id", "name","top_image", "description", "prices", "badges",
```
```suggestion
import java.util.List;

@JsonPropertyOrder({"id", "name","top_image", "description", "prices", "badges",
```

rla36 replied:
아래 댓글과 같은 이유로 `@jsonpropertyorder`을 달았었는데,
현재는 디온이 말씀해주신 방식으로 Json Property naming 설정을 한 후 지웠습니다!

ksundong commented on BE/src/main/java/sidedish/service/dto/DetailDishDTO.java:
```diff
@@ -0,0 +1,110 @@
+import sidedish.service.ConvertUtils;
+
+import java.util.List;
+@JsonPropertyOrder({"id", "name","top_image", "description", "prices", "badges",
+        "stock" , "point", "delivery_info", "thumb_images", "detail_images"})
```
왜 순서를 정의해주셨나요?

rla36 replied:
`@JsonProperty`애노테이션을 추가해서 리스폰스 값들이 스네이크 케이스를 따르도록 설정하니까, 애노테이션을 추가한 값들이 맨 아래로 내려가는 문제가 생겼습니다. ios분들과 합의한 순서에 맞지 않아 순서를 지정해줬습니다. 
리스폰스된 json 값에서 dish의 어떤 속성이 먼저 오는지 순서가 중요한 것 같지는 않았는데, 그래도 합의한 것을 따르는 것이 좋을 것 같아 이렇게 했습니다.

ksundong replied:
생각해보면 순서는 별로 중요하지 않습니다.
key, value 형태이기 때문이에요.

ksundong commented on BE/src/main/java/sidedish/service/dto/DetailDishDTO.java:
```diff
@@ -0,0 +1,110 @@
+    private Long id;
+    private String name;
+
+    @JsonProperty("top_image")
```
이렇게 일일이 정의해주는 것도 괜찮지만,
전체가 공통으로 설정을 공유한다면
Json Parser의 설정을 바꿔보세요. 힌트는 jackson, ObjectMapper입니다~

rla36 replied:
감사합니다! application.properties에 `spring.jackson.property-naming-strategy=SNAKE_CASE` 추가해서 해결했습니다 ㅎㅎ

ksundong commented on BE/src/main/java/sidedish/service/dto/RequestDishDTO.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.List;
+
+@JsonPropertyOrder({"id", "name","top_image", "description", "prices", "badges",
+        "stock" , "point", "delivery_info", "thumb_images", "detail_images", "category"})
```
정렬이 잘 되지 않은 것 같네요.

ksundong commented on BE/src/main/java/sidedish/service/dto/RequestDishDTO.java:
```diff
@@ -0,0 +1,109 @@
+        this.categoryId = categoryId;
+    }
+
+    public Dish createDish() {
+        return new Dish(name, topImage, description, ConvertUtils.convertToString(prices), ConvertUtils.convertToString(badges),
+                stock, deliveryInfo, ConvertUtils.convertToString(thumbImages), ConvertUtils.convertToString(detailImages));
+    }
```
👍

ksundong commented on BE/src/main/resources/application.properties:
```diff
@@ -0,0 +1,10 @@
+logging.level.sql=debug
+
+spring.datasource.initialization-mode=always
+spring.datasource.schema=classpath:schema.sql
+spring.datasource.data=classpath:data.sql
```
기본 설정이 있어서 작성해주지 않으셔도 될겁니다.

rla36 replied:
schema.sql data.sql외의 이름을 사용할 때 저 부분을 application.properties에 추가해주는 것이고, 
기본설정대로 schema.sql, data.sql으로 이름을 정했으면 필요없는거군요!!
처음 알았어요 감사합니당ㅎㅎ

ksundong reviewed (changes requested):
안녕하세요! Jung & Yeon!! 리뷰어 Dion입니다.
두 번째 프로젝트를 진행해주시느라 고생이 많으십니다.
객체와 DB 설계가 조금 아쉬운 부분이 있었어요. (String으로 저장해버리는 경우)
그래도 어느정도 API 설계가 되어있어서 빠르게 수정가능한 부분만 수정해주시면 좋을 것 같습니다.
OOP를 좀 더 시도해봐주시면 좋을 것 같아요.
그리고 예외 처리도 시도해주시면 좋을 것 같습니다.

> RequestDishDTO 객체를 생성하는 메소드(createDish)를 Dish, DishDTO 중 어느 클래스에 넣을지 고민했습니다. 일단 Entity는 서비스, DTO와 별개로 존재해야된다는 점에서 DTO 클래스에 넣는것이 맞을 것 같아 그렇게 했습니다.

어떤 코드는 Entity에 DTO가 드러나고, 어떤 코드는 드러나지 않더라고요.
답을 드리자면, 변경되기 쉬운쪽에 위치하는 것이 맞습니다. Entity보다는 DTO가 변경가능성이 높으니 DTO에 위치하는 것이 좋겠죠?
Entity → DTO 변환, DTO → Entity 변환 모두 DTO 쪽에서 수행해주거나, 중간계층을 거치는 것이 맞겠습니다.

고생하셨습니다~
궁금하신점이나 도움이 필요한 부분, 이해가 되지 않는 부분 등은 코멘트 남겨주세요~

kimnayeon0108 committed:
fix: extend length of TOP_IMAGE column

kimnayeon0108 committed:
fix: change database name

kimnayeon0108 committed:
Add .gitignore

kimnayeon0108 committed:
chore: update .gitignore

kimnayeon0108 committed:
chore: update .gitignore

rla36 committed:
feat: remove @Jsonproperty annotation and add property settings

rla36 committed:
feat: change dish column type from string to list

rla36 committed:
feat: change type of dish property according to DB construction

rla36 committed:
feat: delete unused method and add new TypeConvert methods

rename ConvertUtils to TypeConvertUtils to show the purpose of the class

 # Please enter the commit message for your changes. Lines starting

rla36 committed:
feat: add Exception and handler

add NotFoundException and make DishNotFoundException and
CategoryNotFoundException extend NotFoundException.

add ExceptionHandler to handle NotFoundException

rla36 committed:
chore: fix untracked files

rla36 committed:
feat: change for loop to stream in converting dish to dishDTO

ksundong commented on BE/src/main/java/sidedish/domain/Category.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public boolean hasDish(Dish dish) {
+       if(!dishes.contains(dish)) {
+           throw new DishNotFoundException();
+       } return true;
```
```suggestion
       if (!dishes.contains(dish)) {
           throw new DishNotFoundException();
       }
       return true;
```

ksundong commented on BE/src/main/java/sidedish/exception/CategoryNotFoundException.java:
```diff
@@ -0,0 +1,9 @@
+package sidedish.exception;
+
+public class CategoryNotFoundException extends NotFoundException {
+    private static final String CATEGORY_NOT_FOUND_MESSAGE = "해당 카테고리가 존재하지 않습니다!";
```
잘 추출해주셨네요~

ksundong commented on BE/src/main/java/sidedish/service/DishService.java:
```diff
@@ -0,0 +1,46 @@
+
+    public void saveDishes(List<RequestDishDTO> dishDTOs) {
+        for (RequestDishDTO dishDTO : dishDTOs) {
+            Category category = categoryRepository.findById(dishDTO.getCategoryId()).orElseThrow(CategoryNotFoundException::new);
+            Dish dish = dishDTO.createDish();
+            category.addDish(dish);
+            categoryRepository.save(category);
```
개행을 통해서 논리의 단위를 나눠주셔야 가독성이 높아집니다.

kimnayeon0108 replied:
@ksundong 네엡 변경했습니다. 이부분도 더 신경쓰겠습니다!

ksundong commented on BE/src/main/java/sidedish/service/TypeConvertUtils.java:
```diff
@@ -0,0 +1,65 @@
+        return badges.stream()
+                .map(Badge::new)
+                .collect(Collectors.toList());
+    }
+
+    public static List<ThumbImage> convertToThumbImageList(List<String> images) {
+        return images.stream()
+                .map(ThumbImage::new)
+                .collect(Collectors.toList());
+    }
+
+    public static List<DetailImage> convertToDetailImageList(List<String> images) {
+        return images.stream()
+                .map(DetailImage::new)
+                .collect(Collectors.toList());
+    }
+
+    public static List<Integer> convertPriceList(List<Price> prices) {
+        return prices.stream()
+                .map(Price::getPrice)
+                .collect(Collectors.toList());
+    }
+
+    public static List<String> convertBadgeList(List<Badge> badges) {
+        return badges.stream()
+                .map(Badge::getBadge)
+                .collect(Collectors.toList());
+    }
+
+    public static List<String> convertThumbImageList(List<ThumbImage> thumbImages) {
+        return thumbImages.stream()
+                .map(ThumbImage::getThumbImage)
+                .collect(Collectors.toList());
+    }
+
+    public static List<String> convertDetailImageList(List<DetailImage> detailImages) {
+        return detailImages.stream()
+                .map(DetailImage::getDetailImage)
+                .collect(Collectors.toList());
+    }
```
일급 컬렉션에 대해서 한 번 공부해보세요~

kimnayeon0108 replied:
@ksundong 일급 컬렉션에 대해 학습하고 Prices, Badegs, ThumbImages, DetailImages 클래스를 생성해서 적용해보았습니다!

ksundong commented on BE/src/main/java/sidedish/service/dto/RequestDishDTO.java:
```diff
@@ -0,0 +1,108 @@
+    @JsonProperty("category")
+    private Long categoryId;
+
+    public RequestDishDTO() {
+    }
```
요것도 필요한 시점에 선언해주세요~

kimnayeon0108 replied:
@ksundong 저희가 dish의 모든 데이터를 Post방식으로 한번에 날려서 데이터를 저장하고 있는데, 이때 이부분이 없으면 Cannot construct instance of 'sidedish.service.dtoRequestDishDTO' (no Creators, like default constructor, exist) 에러가 발생해서 선언했습니다!

ksundong replied:
그랬군요. 잘하셨네요!

ksundong commented on BE/src/main/java/sidedish/service/dto/RequestDishDTO.java:
```diff
@@ -0,0 +1,108 @@
+    private List<String> thumbImages;
+    private List<String> detailImages;
+
+    @JsonProperty("category")
+    private Long categoryId;
```
아쉽군요.

kimnayeon0108 replied:
@ksundong 아예 categoryId로 속성값을 지정하는게 더 좋을까요??

ksundong replied:
네 저는 그게 더 낫다고 봅니다.

ksundong reviewed (approved):
고생하셨습니다~ 계속해서 열심히 해주세요!
몇몇 피드백은 반영되진 않았는데, 가능하시다면 반영해서 다음 리뷰 보내주시면 좋겠네요~

rla36 commented:
@ksundong 말씀해주신 부분 수정했습니다! 

추가적으로 추상 클래스를 추가해 중복되는 부분을 제거했습니다.
Dish, DishBuilder클래스의 중복부분은 AbstractDish로, DetailDishDTO, mainPageDishDTO, RequestDishDTO의 중복부분은 AbstractDishDTO로 빼내었습니다.
